From fbf7aec399e3044c6b2bc778a79143ffb283427d Mon Sep 17 00:00:00 2001
From: Shunsuke Kataoka <shunsuke.kataoka.df@renesas.com>
Date: Wed, 26 Jun 2013 14:22:47 +0900
Subject: [PATCH 0679/1083] ASoC: sh: scu: Fix work queue using private worker
 thread

Modify the work queue schedule from the shared queue to the private queue.

Signed-off-by: Shunsuke Kataoka <shunsuke.kataoka.df@renesas.com>
(cherry picked from commit 4ff1f3d2c811fbb9455fd5c6656fcfb5e7723f20)

Signed-off-by: Ryo Kataoka <ryo.kataoka.wt@renesas.com>
---
 include/sound/sh_scu.h | 3 ++-
 sound/soc/sh/scu_pcm.c | 6 ++++--
 2 files changed, 6 insertions(+), 3 deletions(-)

diff --git a/include/sound/sh_scu.h b/include/sound/sh_scu.h
index 73162d3..9a7f382 100644
--- a/include/sound/sh_scu.h
+++ b/include/sound/sh_scu.h
@@ -2128,7 +2128,8 @@ struct scu_pcm_info {
 	spinlock_t pcm_lock;		/* for trigger process */
 	struct dma_chan *de_chan[SHDMA_SLAVE_PCM_MAX];
 	struct sh_dmadesc_slave de_param[SHDMA_SLAVE_PCM_MAX];
-	struct work_struct	work;
+	struct work_struct work;
+	struct workqueue_struct *workq;
 	struct scu_route_info *routeinfo;
 	struct snd_pcm_substream *ss;
 };
diff --git a/sound/soc/sh/scu_pcm.c b/sound/soc/sh/scu_pcm.c
index 53a8328..49c4473 100644
--- a/sound/soc/sh/scu_pcm.c
+++ b/sound/soc/sh/scu_pcm.c
@@ -88,7 +88,7 @@ static void scu_dma_callback(struct snd_pcm_substream *ss)
 	if (pcminfo->flag_start == 0)
 		return;
 
-	schedule_work(&pcminfo->work);
+	queue_work(pcminfo->workq, &pcminfo->work);
 
 	FNC_EXIT
 }
@@ -607,7 +607,7 @@ static int scu_audio_start(struct snd_pcm_substream *ss)
 	/* PCM 1st process */
 	pcminfo->flag_first = 1;
 
-	schedule_work(&pcminfo->work);
+	queue_work(pcminfo->workq, &pcminfo->work);
 
 	FNC_EXIT
 	return ret;
@@ -653,6 +653,7 @@ static struct scu_pcm_info *scu_pcm_new_stream(struct snd_pcm_substream *ss)
 
 	spin_lock_init(&pcminfo->pcm_lock);
 
+	pcminfo->workq = create_workqueue("sh_scu_pcm");
 	INIT_WORK(&pcminfo->work, scu_dma_do_work);
 	FNC_EXIT
 	return pcminfo;
@@ -666,6 +667,7 @@ static void scu_pcm_free_stream(struct snd_pcm_runtime *runtime)
 
 	/* post process */
 	cancel_work_sync(&pcminfo->work);
+	destroy_workqueue(pcminfo->workq);
 	kfree(runtime->private_data);	/* free pcminfo structure */
 
 	FNC_EXIT
-- 
1.8.3.2

